Conversion of JDKPackager to use JDK-provided Ant tasks.

metasim opened the pull request:
A major overhaul, replacing the command-line tool based package generation with an Ant task based one, thereby exposing new capabilities, including:

* Better file mapping support
* JVM arguments
* Application (i.e. `main`) arguments
* System properties
* File associations

The task `jdkpackager:antBuildDefn` can be intercepted, allowing advanced customization via XML DOM processing. `jdkpackager:writeAntBuild` generates the file `target/jdkpackager/build.xml`, which can be inspected for debugging.

muuki88 commented on src/main/scala/com/typesafe/sbt/packager/jdkpackager/JDKPackagerAntHelper.scala:
```diff
@@ -0,0 +1,246 @@
+   *
+   * see: https://docs.oracle.com/javase/8/docs/technotes/guides/deploy/javafx_ant_task_reference.html
+   */
+  private[jdkpackager] def makeAntBuild(
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Method is longer than 50 lines](https://www.codacy.com/app/nepomukseiler/sbt-native-packager/file/637524688/issues/source?bid=2040925&fileBranchId=2311583#l64)

muuki88 replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The number of parameters should not exceed 8](https://www.codacy.com/app/nepomukseiler/sbt-native-packager/file/637524688/issues/source?bid=2040925&fileBranchId=2311583#l64)

metasim replied:
@muuki88 How strict are you looking to be with the coding checks? I think the method "longer than 50 lines" in question is the one containing the Ant XML DOM. I can split it up into multiple functions, but you'd loose the one-to-one correspondence with what ends up getting written to the `build.xml` file. 

Also, in order to reduce the number of parameters to the DOM construction function I'd have to either introduce new intermediate container classes, or add some other level of indirection that I actually think adds complexity. How do you suggest I handle this?

muuki88 replied:
I'll have a closer look. But for good reasons (e.g.  clearly more readable), then rules should be broken.

muuki88 replied:
So. This is a lot of XML inside a code. I know scala supports creating xml. However what do you think, putting this into an `jdkpackager-ant.xml.template` and insert the dynamic parts with place holders?

muuki88 replied:
> Would hate to looses the type checking, which has saved me a number of times. See what you think of the latest update, where I modularized it some.

see https://github.com/sbt/sbt-native-packager/pull/583#discussion_r30338005

muuki88 replied:
Yeah, that looks cleaner.

muuki88 commented on src/main/scala/com/typesafe/sbt/packager/jdkpackager/JDKPackagerPlugin.scala:
```diff
@@ -5,82 +5,78 @@ import com.typesafe.sbt.packager.Keys._
-  object autoImport extends JDKPackagerKeys {
-    val JDKPackager = config("jdkPackager") extend Universal
-  }
+  object autoImport extends JDKPackagerKeys
```
The `autoImport` should also contain the `JDKPackagerToolkit` so the user doesn't have to import it manually.

metasim replied:
Would hate to looses the type checking, which has saved me a number of times. See what you think of the latest update, where I modularized it some.

muuki88 commented on src/main/scala/com/typesafe/sbt/packager/jdkpackager/JDKPackagerPlugin.scala:
```diff
@@ -5,82 +5,78 @@ import com.typesafe.sbt.packager.Keys._
+      packageDescription := packageDescription.value,
+      mappings := (mappings in Universal).value,
+      antPackagerTasks := JDKPackagerAntHelper.locateAntTasks(javaHome.value),
+      antBuildDefn := JDKPackagerAntHelper.makeAntBuild(
```
The number of method parameters is actually fine, as this makes sense here.

muuki88 commented:
One thing regarding the PR. Could you squash the commits into one?

metasim commented:
@muuki88 Not sure I know how to collapse commits. Pointers?

muuki88 commented:
Sure :) In essence you do a `git rebase -i origin/master`, choose the `squash` option and `push -f` your branch. For a bit longer description:

- [Github Tutorial](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit)
- [Other Tutorial](http://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit)

metasim commented:
@muuki88 Thanks for your patience with this one. Glad to know how to `squash` now! :-)

muuki88 commented:
No problem :) Always happy to help.

muuki88 commented:
In the test project on windows I get

```
[error] (jdkPackager:packageBin) E:\sbt-native-packager\test-project-jdkpackager\target\jdkpackager\build.xml:25: fx:info doesn't support the nested "association" element.
```

On Ubuntu 15.04 with `java version "1.8.0_45` I get this

```
[error] (jdkPackager:antBuildDefn) Please set key `antPackagerTasks in JDKPackager` to `ant-javafx.jar` path, which should be find in the `lib` directory of JDK 8 installation.
[error] Total time: 7 s, completed 15.05.2015 22:44:22
```

metasim commented:
Ooof. Sorry about that. Thanks for the extra check. Will address soon.

Wonder why Travis CI didn't catch the Linux one?

muuki88 commented:
Yeah, that's weird as travis uses the oracle jdk as well. However they have ubuntu 14.04 LTS I guess.

muuki88 commented:
This seems to be an Ubuntu specific problem. The old jdk-packager version doesn't work as well.

muuki88 commented on src/sphinx/formats/jdkpackager.rst:
```diff
@@ -136,7 +152,19 @@ Here's what the build file looks like:
+        FileAssociation("foobar", "application/foobar", "Foobar file type"),
+        FileAssociation("barbaz", "application/barbaz", "Barbaz file type", jdkAppIcon.value)
+    )
 
```
Could you add one small section here, showing or to add other jdk-locations (if not found by default), like this

```scala
jdkPackagerTool := jdkPackagerTool.value orElse Some(file("/usr/lib/jvm/java-8-oracle/bin/javapackager"))
```

and how setting `JDK_HOME` or `JAVA_HOME` may help ;)

metasim replied:
Will do!

muuki88 commented:
one last comment. I think a small documentation hint should do the trick as jvm detection is already sophisticated and we can make it more sophisticated in follow up pull requests.

metasim commented:
@muuki88 On your last comment, do you mean additional info in the Sphinx docs, or do you mean in the error message?

muuki88 commented:
@metasim I meant sphinx. I think the error message is explanatory enough

metasim commented:
@muuki88  Regarding this:

> In the test project on windows I get

>     [error] (jdkPackager:packageBin) E:\sbt-native-packager\test-project-jdkpackager\target\jdkpackager\build.xml:25: fx:info doesn't support the nested "association" element.

What version of Java were you using? I'm wondering if the file association feature was added in JDK 8 build 40 (yes... Oracle added major new features in a build-increment release).

muuki88 commented:
I think I have an older one. Will try again, but as this is still experimental and working on your machine, we will refine later on.

metasim commented:
@muuki88 In addition to some updates to the documentation, I added debug mode logging of the search process for `ant-javafx.jar` (which replaces searching for `javapackager`). Also added settings key docs on file associations needing JDK 8 build >= 40.

muuki88 commented:
brilliant. can you squash again and I'll merge it right away :)

metasim commented:
@muuki88 Squashed. :-)

metasim commented:
@muuki88 Just so you're aware (as the project lead), this plugin now depends on the `LauncherJarPlugin`, depending upon the `Main-Class` and `Class-Path` manifest attributes defined therein. I've tried to rely on the by-products of the other packaging plugins whenever I can.

muuki88 commented:
Awesome work ( as always ). Thanks for highlighting the dependency. The classpath plugins are getting used more and more, so using these is totally fine.

muuki88 commented:
I'll make an release ASAP

metasim commented:
@muuki88 Thanks so much. Don't forget to release the docs too (don't know if it's automatic).... old ones would really throw off users.

File associations should be a winning feature for some.
